Migrate op to new infra: data_movement/transpose

shutovilyaep opened the pull request:
Migration completed:
- [x] Created operation_attributes_t struct with dim, output_mem_config, pad_value
- [x] Created tensor_args_t struct with input tensor reference
- [x] Defined tensor_return_value_t and spec_return_value_t
- [x] Implemented compute_output_specs
- [x] Implemented create_output_tensors
- [x] Implemented select_program_factory returning variant of 6 factories
- [x] Implemented validate_on_program_cache_miss and validate_on_program_cache_hit
- [x] Registered ttnn::prim::transpose
- [x] Updated transpose.cpp to use ttnn::prim::transpose instead of operation::run
- [x] Created 6 program factories with shared_variables_t, create, override_runtime_arguments
- [x] Removed old Transpose struct and transpose_op.hpp/.cpp

Program factories split into separate files:
- transpose_cn_program_factory.hpp/.cpp
- transpose_hc_program_factory.hpp/.cpp
- transpose_hc_sharded_program_factory.hpp/.cpp
- transpose_wh_program_factory.hpp/.cpp
- transpose_wh_sharded_program_factory.hpp/.cpp
- transpose_wh_sharded_rm_program_factory.hpp/.cpp

Resolves: #32761

### Ticket
Link to Github Issue

### Problem description
Provide context for the problem.

### What's changed
Describe the approach used to solve the problem.
Summarize the changes made and its impact.

### Checklist
- [X] One more attempt to run CI https://github.com/tenstorrent/tt-metal/actions/runs/20071473367
- [X] Code analysis https://github.com/tenstorrent/tt-metal/actions/runs/20071476137

shutovilyaep committed:
Migrate data_movement/transpose device operation to TMP pattern

Migration completed:
- [x] Created operation_attributes_t struct with dim, output_mem_config, pad_value
- [x] Created tensor_args_t struct with input tensor reference
- [x] Defined tensor_return_value_t and spec_return_value_t
- [x] Implemented compute_output_specs
- [x] Implemented create_output_tensors
- [x] Implemented select_program_factory returning variant of 6 factories
- [x] Implemented validate_on_program_cache_miss and validate_on_program_cache_hit
- [x] Registered ttnn::prim::transpose
- [x] Updated transpose.cpp to use ttnn::prim::transpose instead of operation::run
- [x] Created 6 program factories with shared_variables_t, create, override_runtime_arguments
- [x] Removed old Transpose struct and transpose_op.hpp/.cpp

Program factories split into separate files:
- transpose_cn_program_factory.hpp/.cpp
- transpose_hc_program_factory.hpp/.cpp
- transpose_hc_sharded_program_factory.hpp/.cpp
- transpose_wh_program_factory.hpp/.cpp
- transpose_wh_sharded_program_factory.hpp/.cpp
- transpose_wh_sharded_rm_program_factory.hpp/.cpp

Resolves: #32761

shutovilyaep commented:
CI script ran https://github.com/tenstorrent/tt-metal/actions/runs/19772002008 [RUNNING for now]

shutovilyaep committed:
fix: auto*

shutovilyaep commented:
APC CI script https://github.com/tenstorrent/tt-metal/actions/runs/19867326809 [RUNNING]
```json
{"sd-unit-tests":false,"fast-dispatch-unit-tests":false,"fabric-unit-tests":false,"cpp-unit-tests":false,"ttnn-unit-tests":true,"models-unit-tests":false,"tt-train-cpp-unit-tests":false,"run-profiler-regression":false,"t3000-apc-fast-tests":false,"test-ttnn-tutorials":false,"triage-tests":false}
```

aliaksei-sala commented on ttnn/cpp/ttnn/operations/data_movement/transpose/device/transpose_device_operation.cpp:
```diff
@@ -127,16 +167,16 @@ void Transpose::validate(const std::vector<Tensor>& input_tensors) const {
+    const auto& dim = operation_attributes.dim;
+    const auto& output_mem_config = operation_attributes.output_mem_config;
 
-    // TODO: Remove usage of input/output padded shape
```
Why do we remove TODO ?

aliaksei-sala commented on ttnn/cpp/ttnn/operations/data_movement/transpose/transpose.cpp:
```diff
@@ -60,13 +58,13 @@ inline Tensor transpose_(
 
 ttnn::Tensor transpose_nd(
     const ttnn::Tensor& input_tensor,
-    const uint32_t dim1,
```
Why remove const ?

aliaksei-sala commented on ttnn/cpp/ttnn/operations/data_movement/transpose/device/transpose_program_factory.cpp:
```diff
@@ -1,2084 +0,0 @@
-            num_non_repeat_cores = unique_values.size();
-            read_stick_stride = read_stick_offset[1] - read_stick_offset[0];
-
-            // TODO: add the second batch args (num_non_repeat_cores, read_stick_offset, non_repeat_noc_x_values,
```
TODO was removed in a new code

aliaksei-sala commented on ttnn/cpp/ttnn/operations/data_movement/transpose/device/transpose_program_factory.cpp:
```diff
@@ -1,2084 +0,0 @@
-    uint32_t num_cores_x = grid_size.x;
-    uint32_t num_cores_y = grid_size.y;
-
-    log_debug(tt::LogOp, "all_cores: {}", all_cores);
```
log_debug messages were removed in a new code

aliaksei-sala commented on ttnn/cpp/ttnn/operations/data_movement/transpose/device/transpose_program_factory.cpp:
```diff
@@ -1,2084 +0,0 @@
-
-operation::ProgramWithCallbacks transpose_hc_multi_core_tiled_interleaved(
-    const Tensor& a, Tensor& output, const std::optional<float>& pad_value) {
-    TT_ASSERT(a.storage_type() == StorageType::DEVICE, "Operand to transpose_hc needs to be on device!");
```
Both TT_ASERT were removed  in a new code

aliaksei-sala commented on ttnn/cpp/ttnn/operations/data_movement/transpose/device/transpose_program_factory.cpp:
```diff
@@ -1,2084 +0,0 @@
-    tt::DataFormat cb_data_format = tt::tt_metal::datatype_to_dataformat_converter(a.dtype());
-    uint32_t single_tile_size = tt::tile_size(cb_data_format);
-
-    log_debug(tt::LogOp, "transpose_hc_multi_core");
```
log_debug, TODO comments and TT_ASSERT were removed in a new code
I believe this is the common problem across all factories
please verify that all missing parts have been migrated.

aliaksei-sala commented on ttnn/cpp/ttnn/operations/data_movement/transpose/device/transpose_hc_program_factory.cpp:
```diff
@@ -0,0 +1,601 @@
+
+    bool is_tiled_interleaved = input_tensor.layout() == Layout::TILE && !input_tensor.is_sharded();
+
+    if (is_tiled_interleaved) {
```
Let's create one factory per program
TransposeHCProgramFactory can be split into two separate factories.

shutovilyaep committed:
fix: initializers, clang-tidy

shutovilyaep committed:
fix: namespace data_movement::transpose

shutovilyaep committed:
fix: Tensor in tensor_args_t

shutovilyaep committed:
fix: create_op_performance_model

shutovilyaep committed:
fix: Split TransposeHCProgramFactory into separate factories

- Add get_parallelization_strategy function, use it in select_program_factory
- Split TransposeHCProgramFactory into 3 separate factories:
  * TransposeHCTiledInterleavedProgramFactory (tiled interleaved with padding)
  * TransposeHCTiledProgramFactory (tiled non-interleaved)
  * TransposeHCRMProgramFactory (row-major)
- Remove umbrella header, include factory headers directly
- Update CMakeLists.txt with new source files

shutovilyaep committed:
fix: logging, TODOs as before TMP migration

shutovilyaep commented on ttnn/cpp/ttnn/operations/data_movement/transpose/device/transpose_wh_sharded_rm_program_factory.hpp:
```diff
@@ -0,0 +1,38 @@
+
+#include <tt-metalium/core_coord.hpp>
+
+namespace ttnn::operations::data_movement::program {
```
fixed: added transpose-related files to ::transpose namespace

shutovilyaep commented on ttnn/cpp/ttnn/operations/data_movement/transpose/device/transpose_device_operation.cpp:
```diff
@@ -1,38 +1,74 @@
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    const auto& input_tensor = tensor_args.input;
+    bool is_sharded = input_tensor.is_sharded();
+    bool is_l1 = is_sharded && input_tensor.buffer()->is_l1();
```
fixed: using get_parallelization_strategy

shutovilyaep committed:
fix: added asserts to factories

shutovilyaep committed:
fix: debug logs

shutovilyaep commented:
All pre-commits CI run: https://github.com/tenstorrent/tt-metal/actions/runs/19906755194

shutovilyaep committed:
fix: missed TT_ASSERT after the refactoring

shutovilyaep committed:
fix: comments

shutovilyaep committed:
fix: removed set_or_update_runtime_arguments

shutovilyaep committed:
Fix: Initialize TransposeOpDim enum in operation_attributes_t struct

- Add {} default initialization to dim member to fix clang-tidy warning
- MemoryConfig and std::optional members already have proper defaults
